### PR TITLE
[Merged by Bors] - feat: noncommutative coproduct for MonoidHom

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
Add `MonoidHom.noncommCoprod` which allows to define a morphism from a product `M x N` 
to a monoid `P` given two monoid morphisms `M -> P` and `N -> P`,
provided a commutativity assumption on the ranges.

---

- [x] depends on: #8756

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Prod.lean [hunk 1]
```diff
@@ -20,7 +21,9 @@ trivial `simp` lemmas, and define the following operations on `MonoidHom`s:
 * `inl M N : M →* M × N`, `inr M N : N →* M × N`: inclusions of first/second monoid
   into the product;
 * `f.prod g` : `M →* N × P`: sends `x` to `(f x, g x)`;
-* `f.coprod g : M × N →* P`: sends `(x, y)` to `f x * g y`;
+* (when `P` is commutative) `f.coprod g : M × N →* P` sends `(x, y)` to `f x * g y`,
+and `comm` says that the elements in the ranges of `f` and `g` commute pairwise,
+`f.noncommCoprod g comm` is the analogous `M × N →* P`.
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean:26]:
```suggestion
* (when `P` is commutative) `f.coprod g : M × N →* P` sends `(x, y)` to `f x * g y`,
  and `comm` says that the elements in the ranges of `f` and `g` commute pairwise,
  `f.noncommCoprod g comm` is the analogous `M × N →* P`.
```

File: Mathlib/Algebra/Group/Prod.lean [hunk 2]
```diff
@@ -434,6 +439,51 @@ theorem comp_coprod {Q : Type*} [CommSemigroup Q] (h : P →ₙ* Q) (f : M →
...
+/-- Coproduct of two `MulHom`s with the same codomain with commutation assumption :
+  `f.noncommCoprod g _ (p : M × N) = f p.1 * g p.2`.
+  (For the commutative case, use `MulHom.coprod`) -/
+@[to_additive
+    "Coproduct of two `AddHom`s with the same codomain with commutation assumption:
+    `f.noncommCoprod g _ (p : M × N) = f p.1 + g p.2`.
+    (For the commutative case, use `AddHom.coprod`)"]
+def noncommCoprod : M × N →ₙ* P where
+  toFun := fun mn ↦ (f mn.fst) * (g mn.snd)
+  map_mul' := fun mn mn' ↦ by
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean:456]:
```suggestion
  toFun := fun mn ↦ (f mn.fst) * (g mn.snd)
  map_mul' := fun mn mn' ↦ by
```
```suggestion
  toFun mn := (f mn.fst) * (g mn.snd)
  map_mul' mn mn' := by
```

File: Mathlib/Algebra/Group/Prod.lean [hunk 3]
```diff
@@ -447,43 +447,22 @@ variable {M N P : Type*} [Mul M] [Mul N] [Semigroup P]
...
     (For the commutative case, use `AddHom.coprod`)"]
 def noncommCoprod : M × N →ₙ* P where
-  toFun mn := (f mn.fst) * (g mn.snd)
-  map_mul' mn mn' := by
-    simp only [Prod.fst_mul, Prod.snd_mul, map_mul]
-    simp only [mul_assoc]
-    apply congr_arg₂ _ rfl
-    simp only [← mul_assoc]
-    apply congr_arg₂ _ _ rfl
-    apply comm
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean]: The trick here is knowing that `mul_mul_mul_comm` exists!

File: Mathlib/Algebra/Group/Prod.lean [hunk 4]
```diff
@@ -447,43 +447,22 @@ variable {M N P : Type*} [Mul M] [Mul N] [Semigroup P]
...
-  rfl
-
-/- -- simp_normal_form complains
-@[to_additive (attr := simp)]
-theorem noncommCoprod_apply' (p : M × N) :
-    f.noncommCoprod g comm p = g p.2 * f p.1 :=
-  comm p.1 p.2
--/
+  toFun mn := f mn.fst * g mn.snd
+  map_mul' mn mn' := by simpa using (comm _ _).mul_mul_mul_comm _ _
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean:457]: We shouldn't be committing code in comments; though perhaps this was just you asking for feedback on why `simpNF` complains.

I think this lemma is probably fine to have without the `simp` attribute. You might also want to add the one that says how `noncommCoprod` composes with whatever the name of the multiplicative map `(x, y) => (y, x)` is.

File: Mathlib/Algebra/Group/Prod.lean [hunk 5]
```diff
@@ -447,43 +447,22 @@ variable {M N P : Type*} [Mul M] [Mul N] [Semigroup P]
...
 
 @[to_additive]
 theorem comp_noncommCoprod {Q : Type*} [Semigroup Q] (h : P →ₙ* Q) :
     h.comp (f.noncommCoprod g comm) =
-      (h.comp f).noncommCoprod (h.comp g) (fun m n ↦ by
-        simp only [coe_comp, Function.comp_apply]
-        exact Commute.map (comm m n) h) :=
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean]: The trick here is to note that the `simp only` line looks like obvious definitional equalities, so can be removed.

File: Mathlib/Algebra/Group/Prod.lean [hunk 6]
```diff
@@ -737,69 +716,39 @@ variable {P : Type*} [Monoid P]
...
-  rfl
-
-/- -- simp_normal_form complains
-@[to_additive (attr := simp)]
-theorem noncommCoprod_apply' (p : M × N) :
-    f.noncommCoprod g comm p = g p.2 * f p.1 :=
-  comm p.1 p.2
--/
+  map_one' := by simp only [Prod.fst_one, Prod.snd_one, map_one, mul_one]
+  __ := f.toMulHom.noncommCoprod g.toMulHom comm
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean:727]: This one is a pretty important trick; don't reprove the same thing again if you can re-use the existing structure from the `MulHom`.

File: Mathlib/Algebra/Group/Prod.lean [hunk 7]
```diff
@@ -737,69 +716,39 @@ variable {P : Type*} [Monoid P]
...
 @[to_additive (attr := simp)]
 theorem noncommCoprod_comp_inr : (f.noncommCoprod g comm).comp (inr M N) = g :=
-  ext fun x => by simp [noncommCoprod_apply]
-
-@[to_additive]
-theorem noncommCoprod.commute (f : M × N →* P) (m : M) (n : N) :
-    Commute (f.comp (inl M N) m) (f.comp (inr M N) n) := by
-  simp only [coe_comp, Function.comp_apply]
-  apply Commute.map
-  apply inl_inr_commute
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean]: I removed this lemma, as the name was a bit strange, the lemma isn't in simpNF (if it were, the `comp` would be eliminated) and `(inl_inr_commute _ _).map f` is a pretty short proof.

File: Mathlib/Algebra/Group/Prod.lean [hunk 8]
```diff
@@ -737,69 +716,39 @@ variable {P : Type*} [Monoid P]
...
-theorem noncommCoprod_apply (p : M × N) :
-  f.noncommCoprod g comm p = f p.1 * g p.2 :=
-  rfl
-
-/- -- simp_normal_form complains
-@[to_additive (attr := simp)]
-theorem noncommCoprod_apply' (p : M × N) :
-    f.noncommCoprod g comm p = g p.2 * f p.1 :=
-  comm p.1 p.2
--/
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean]: Feel free to add this one back (not in a comment) if you want to keep it.

**Comment by riccardobrasca** [comment]
This is exactly the same as [this](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/NoncommPiCoprod.html) file but for the product of two things rather than the Pi type, right?

I don't know if it is a good idea to move everything to `NoncommPiCoprod`, but if we keep the two constructions in different files each doc should mention the other file.

File: Mathlib/Algebra/Group/Prod.lean (hunk 4 above)
Comment (AntoineChambert-Loir) [reply Mathlib/Algebra/Group/Prod.lean:457]: I think I can remove it.
What you suggest would be obtained using MulHom.comp, so I'm not sure it is necessary to add it.
(And I don't use noncommProd anymore now…)

**Comment by AntoineChambert-Loir** [comment]
Yes, Riccardo, you're right.

**Comment by AntoineChambert-Loir** [comment]
I added the comment about `MonoidHom.noncommPiCoprod`.

File: Mathlib/Algebra/Group/Prod.lean [hunk 9]
```diff
@@ -410,10 +414,12 @@ section Coprod
...
 /-- Coproduct of two `MulHom`s with the same codomain:
-`f.coprod g (p : M × N) = f p.1 * g p.2`. -/
+  `f.coprod g (p : M × N) = f p.1 * g p.2`.
+  (Commutative codomain; for the general case, see `MulHom.noncommCoprod`) -/
 @[to_additive
-      "Coproduct of two `AddHom`s with the same codomain:
-      `f.coprod g (p : M × N) = f p.1 + g p.2`."]
+    "Coproduct of two `AddHom`s with the same codomain:
+    `f.coprod g (p : M × N) = f p.1 + g p.2`.
+    (Commutative codomain; for the general case, see `MulHom.noncommCoprod`)"]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean:422]:
```suggestion
    (Commutative codomain; for the general case, see AddHom.noncommCoprod`)"]
```

**Comment by eric-wieser** [comment]
I think either this content should move to `GroupTheory/NoncommPiCoprod.lean`, or it should go in a new file `GroupTheory/NoncommCoprod.lean`. The advantage of the new file is that it's very easy to compare the API between the `pi` and `prod` cases.

**Comment by riccardobrasca** [comment]
I agree with Eric about a new file, otherwise LGTM.

**Comment by AntoineChambert-Loir** [comment]
I'll do that soon, but I need some rest now…

File: Mathlib/GroupTheory/NoncommCoprod.lean [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+-/
+import Mathlib.Algebra.Group.Prod
+import Mathlib.Algebra.Hom.Commute
+
+/-!
+# Canonical homomorphism from a pair of monoids
+
+This file defines the construction of the canonical homomorphism from a pair of monoids.
+
+TODO: write the rest of this docstrnig copying (and/or referenceing) `NoncommPiCoprod`
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/NoncommCoprod.lean:14]: I split the file for you, but this docstring still needs writing

**Comment by eric-wieser** [comment]
maintainer merge

I think this is probably good now, though someone should take a final sweep

**Comment by AntoineChambert-Loir** [comment]
I presume this “someone” must be someone else than me?
Antoine 

⁣--
Université Paris Cité 
Institut de mathématiques de Jussieu Paris Rive Gauche
Bâtiment Sophie Germain, 8 place Aurélie Nemours, 75205 Paris cedex 13

Envoyé depuis mon téléphone portable, pardonnez mon éventuelle brièveté​

Le 14 déc. 2023 à 19:17, à 19:17, Eric Wieser ***@***.***> a écrit:

File: Mathlib/Algebra/Group/Prod.lean [hunk 11]
```diff
@@ -12,15 +12,18 @@ import Mathlib.Algebra.GroupWithZero.Units.Basic
...
+We also prove trivial `simp` lemmas, and define the following operations on `MonoidHom`s:
 
 * `fst M N : M × N →* M`, `snd M N : M × N →* N`: projections `Prod.fst` and `Prod.snd`
   as `MonoidHom`s;
 * `inl M N : M →* M × N`, `inr M N : N →* M × N`: inclusions of first/second monoid
   into the product;
 * `f.prod g` : `M →* N × P`: sends `x` to `(f x, g x)`;
-* `f.coprod g : M × N →* P`: sends `(x, y)` to `f x * g y`;
+* (when `P` is commutative) `f.coprod g : M × N →* P` sends `(x, y)` to `f x * g y`,
+  and `comm` says that the elements in the ranges of `f` and `g` commute pairwise,
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Group/Prod.lean:24]:
```suggestion
  and if `comm` says that the elements in the ranges of `f` and `g` commute pairwise,
```
I have the impression the sentence is a little unclear.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/Algebra/Group/Prod.lean (hunk 11 above)
Comment (eric-wieser) [reply Mathlib/Algebra/Group/Prod.lean:24]: I'm not even sure what `comm` refers to here; is this intended for the docstring in the other file?

Comment (riccardobrasca) [reply Mathlib/Algebra/Group/Prod.lean:24]: It is the name of the assumption. Anyway we agree that it is not clear.

Comment (eric-wieser) [reply Mathlib/Algebra/Group/Prod.lean:24]: `coprod` has no such assumption.

Comment (AntoineChambert-Loir) [reply Mathlib/Algebra/Group/Prod.lean]: I made a small change in the formulation by adding a new item for the noncommProd,
I believe it is clearer now.

File: Mathlib/Algebra/Group/Prod.lean [hunk 12]
```diff
@@ -12,15 +12,18 @@ import Mathlib.Algebra.GroupWithZero.Units.Basic
...
+We also prove trivial `simp` lemmas, and define the following operations on `MonoidHom`s:
 
 * `fst M N : M × N →* M`, `snd M N : M × N →* N`: projections `Prod.fst` and `Prod.snd`
   as `MonoidHom`s;
 * `inl M N : M →* M × N`, `inr M N : N →* M × N`: inclusions of first/second monoid
   into the product;
 * `f.prod g` : `M →* N × P`: sends `x` to `(f x, g x)`;
-* `f.coprod g : M × N →* P`: sends `(x, y)` to `f x * g y`;
+* (when `P` is commutative) `f.coprod g : M × N →* P` sends `(x, y)` to `f x * g y`;
+* `f.noncommCoprod g comm` is the analogous `M × N →* P`,
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Prod.lean:24]: This is in a list of things defined in this file, but is not defined in this file!

**Pull request closed without merging**